app/src/main/java/org/mozilla/firefox/vpn/receivers/Boot.kt

```diff
@@ -0,0 +1,63 @@
+/* This Source Code Form is subject to the terms of the Mozilla Public
+ * License, v. 2.0. If a copy of the MPL was not distributed with this
+ * file, You can obtain one at https://mozilla.org/MPL/2.0/. */
+
+package org.mozilla.firefox.vpn.receivers
+
+import android.content.BroadcastReceiver
+import android.content.Context
+import android.content.Intent
+import android.os.Build
+import android.util.Log
+import kotlinx.coroutines.CoroutineScope
+import kotlinx.coroutines.Dispatchers
+import kotlinx.coroutines.launch
+import org.mozilla.firefox.vpn.ext.connectVpnIfPossible
+
+/**
+ * Boot broadcast receiver for auto-starting VPN service on pre-N devices.
+ * This receiver is disabled for N+ devices. On these, Always-On VPN mode exists at the system-level,
+ * and is the system-approved way of doing this.
+ * In fact, newer Androids won't even allow us to do background work (such as starting a service) on boot.
+ */
+class LegacyBootBroadcastReceiver : SafeBootBroadcastReceiver(Intent.ACTION_BOOT_COMPLETED) {
+    override val enabled = Build.VERSION.SDK_INT < Build.VERSION_CODES.N
+    override val logTag = "LegacyBootBroadcastReceiver"
+}
+
+abstract class SafeBootBroadcastReceiver(private val action: String) : BroadcastReceiver() {
```

Comment:
nit: This feels a bit over-engineered. But I am lacking context here, do we expect to implement more such receivers?

Reply:
Ah, you're right - I didn't realize "always-on" was a thing, and had a receiver for handling direct boot intents on newer devices, etc...

So there _were_ two receivers :)